dialogue-core/src/main/java/com/palantir/dialogue/core/BalancedNodeSelectionStrategyChannel.java

```diff
@@ -342,4 +436,105 @@ public void markRequestMade() {
     }
+
+    @VisibleForTesting
+    enum RttEndpoint implements Endpoint {
+        INSTANCE;
+
+        @Override
+        public void renderPath(Map<String, String> _params, UrlBuilder _url) {}
+
+        @Override
+        public HttpMethod httpMethod() {
+            return HttpMethod.OPTIONS;
+        }
+
+        @Override
+        public String serviceName() {
+            return "Balanced";
+        }
+
+        @Override
+        public String endpointName() {
+            return "rtt";
+        }
+
+        @Override
+        public String version() {
+            return "0.0.0";
+        }
+    }
+
+    /**
+     * Always returns the *minimum* value from the last few samples, so that we exclude slow calls that might include
+     * TLS handshakes.
+     */
+    @VisibleForTesting
+    @ThreadSafe
+    static final class RttMeasurement {
+        private static final int NUM_MEASUREMENTS = 5;
+
+        private final long[] samples;
+        private volatile long bestRttNanos = Long.MAX_VALUE;
+
+        RttMeasurement() {
+            samples = new long[NUM_MEASUREMENTS];
+            Arrays.fill(samples, Long.MAX_VALUE);
+        }
+
+        long getNanos() {
+            return bestRttNanos;
+        }
+
+        synchronized void addMeasurement(long newMeasurement) {
+            Preconditions.checkArgument(newMeasurement > 0, "Must be greater than zero");
+            Preconditions.checkArgument(newMeasurement < Long.MAX_VALUE, "Must be less than MAX_VALUE");
+
+            if (samples[0] == Long.MAX_VALUE) {
+                Arrays.fill(samples, newMeasurement);
+                bestRttNanos = newMeasurement;
+            } else {
+                System.arraycopy(samples, 1, samples, 0, NUM_MEASUREMENTS - 1);
+                samples[NUM_MEASUREMENTS - 1] = newMeasurement;
+                bestRttNanos = Arrays.stream(samples).min().getAsLong();
+            }
+        }
+
+        @Override
+        public String toString() {
+            return "RttMeasurement{" + "bestRttNanos=" + bestRttNanos + ", samples=" + Arrays.toString(samples) + '}';
+        }
+    }
+
+    private static final class RttMeasurementRateLimiter {
+        private static final long BETWEEN_SAMPLES = Duration.ofSeconds(1).toNanos();
+
+        private final Ticker clock;
+        private final AtomicLong lastMeasured = new AtomicLong(0);
+        private final AtomicBoolean currentlySampling = new AtomicBoolean(false);
+
+        private RttMeasurementRateLimiter(Ticker clock) {
+            this.clock = clock;
+        }
+
+        Optional<RttMeasurementPermit> acquireNonBlocking() {
+            long last = lastMeasured.get();
+            if (last + BETWEEN_SAMPLES > clock.read()) {
+                return Optional.empty();
+            }
+
+            if (currentlySampling.compareAndSet(false, true)) {
+                lastMeasured.set(clock.read());
+                return Optional.of(() -> currentlySampling.compareAndSet(true, false));
```

Comment:
No need to check the current value, any time this executes the resulting value is `false`.
The value returned here could be a singleton.
```suggestion
                return Optional.of(() -> currentlySampling.set(false));
```